### ISSUE TITLE
JClouds 2.4.0 -> s3proxy 1.9.0 causes Sig errors on s3proxy

### ISSUE THREAD

**Issue by cstamas** [gaul/s3proxy#392]
Using JClouds 2.4.0 S3 API against S3Proxy 1.9.0 (using transient, but tried with file on Linux FS supporting attributes storage)

For one of the request, we get signature errors (so s3proxy claims jclouds provided AWS signature is wrong). while many other request is just fine.

Note: we get this problem with any combo that uses 1.9.0 s3proxy. This is a test suite, and all (many many similar) other S3 requests works okay, but one (and only one!) fails with 1.9.0 s3proxy. We ran the test suite with these combos:
* jclouds 2.3.0 + s3proxy 1.8.0 => PASSES
* jclouds 2.3.0 + s3proxy 1.9.0 => FAILS
* jclouds 2.4.0 + s3proxy 1.8.0 => PASSES
* jclouds 2.4.0 + s3proxy 1.9.0 => FAILS

On "client" (jclouds) side we see nothing special, except it retries for 5 times and gives up:
https://gist.github.com/cstamas/d4aad579480a3217fb1738c756ef77ae

On "server" (s3proxy) side we see errors about signature 5 times (excerpt):
https://gist.github.com/cstamas/9d5e959274bfe46d4803dad93886ac06

Again, this happens for ONE out of many request, this one below PASS OK w/ s3proxy 1.9.0:
```
PUT //127.0.0.1:39941/px01-bkt-0001/content/countall-h/8d/8dda82998fe00f3218b09833b3c0d53e43f5bbe2/17d56acc54800d3 HTTP/1.1
Expect: 100-continue
x-amz-meta-uid: countAll-h/package/hellouniverse/0.0.5/hellouniverse.0.0.5.numeta
x-amz-meta-contentkey: px01-bkt-0001/content/countall-h/8d/8dda82998fe00f3218b09833b3c0d53e43f5bbe2/17d56acc54800d3
x-amz-meta-pxm1: {"attr":{"storageItem-path":"/package/hellouniverse/0.0.5/hellouniverse.0.0.5.numeta","storageItem-readable":"true","storageItem-writable":"true","storageItem-created":"1637836703045","storageItem-modified":"1637836703045","storageItem-l
astRequested":"1637836703045","storageItem-checkedRemotely":"1637836703045","storageItem-storedLocally":"1637836703045","storageItem-repositoryId":"countAll-h","contentGenerator":"interpolate","nuget.mimeType":"application/atom+xml; charset=utf-8; type=p
ackages","storageItem-expired":"false","request.user":"admin","request.address":"127.0.0.1","request.agent":"Nexus-Client/unknown","storageItem-generation":"1","digest.sha1":"fbd1e45db5424a33bc2fdb5c39da2082bb0bee56","digest.md5":"a3c64605398d71f7200b72b
6c6f99eaa","digest.sha256":"8a99c402a39f744703e2c48099a8a92a4f994323860fa6b585642eb3145c73dc","digest.sha512":"d07740710a5c587ba5f0ab7a5aa545073f123ad4394c04037197cb7f36a088a4498201516565e4d3f7ad73f3fb5e6fbdb172121411888a417f4befa87ff99440"},"localStorag
e-itemType":"file","localStorage-contentType":"application/atom+xml; charset=utf-8; type=packages","localStorage-contentSize":"3116","localStorage-contentId":"px01-bkt-0001/content/countall-h/8d/8dda82998fe00f3218b09833b3c0d53e43f5bbe2/17d56acc54800d3"}
x-amz-meta-pointerkey: px01-bkt-0001/pointer/countall-h/8d/8dda82998fe00f3218b09833b3c0d53e43f5bbe2
Date: Thu, 25 Nov 2021 10:37:43 GMT
Authorization: AWS s3proxy:s18zFhCS+amw7H9Hbdpe3ZtbqnE=
User-Agent: jclouds/2.4.0 httpclient/4.1.1
Content-Length: 3116
Content-Type: application/atom+xml; charset=utf-8; type=packages
Host: 127.0.0.1:39941
Connection: keep-alive
Accept-Encoding: gzip,deflate
```
while this one (and ONLY this one) FAILS with s3proxy 1.9.0:
```
PUT //127.0.0.1:39941/px01-bkt-0001/content/countall/e6/e6e29e35a58d31b050dca8aadbbae9d93cb0fe55/17d56aced5601c3 HTTP/1.1
Expect: 100-continue
x-amz-meta-uid: countAll/.metadata/Search/$count/c0/c0182aee291cdaf4574a42d0ac146ccfa79596e0.xml
x-amz-meta-contentkey: px01-bkt-0001/content/countall/e6/e6e29e35a58d31b050dca8aadbbae9d93cb0fe55/17d56aced5601c3
x-amz-meta-pxm1: {"attr":{"storageItem-path":"/.metadata/Search/$count/c0/c0182aee291cdaf4574a42d0ac146ccfa79596e0.xml","storageItem-readable":"true","storageItem-writable":"true","storageItem-created":"1637836713291","storageItem-modified":"163783671329
1","storageItem-lastRequested":"1637836713291","storageItem-checkedRemotely":"1637836713275","storageItem-storedLocally":"1637836713291","storageItem-repositoryId":"countAll","storageItem-expired":"false","proxyRepository-invalidationToken":"121350917965
38","nuget.mimeType":"text/plain;charset=utf-8","request.user":"admin","request.address":"127.0.0.1","request.agent":"Nexus-Client/unknown","storageItem-generation":"1","digest.sha1":"77de68daecd823babbb58edb1c8e14d7106e83bb","digest.md5":"eccbc87e4b5ce2
fe28308fd9f2a7baf3","digest.sha256":"4e07408562bedb8b60ce05c1decfe3ad16b72230967de01f640b7e4729b49fce","digest.sha512":"3bafbf08882a2d10133093a1b8433f50563b93c14acd05b79028eb1d12799027241450980651994501423a66c276ae26c43b739bc65c4e16b10c3af6c202aebb"},"lo
calStorage-itemType":"file","localStorage-contentType":"text/plain;charset=utf-8","localStorage-contentSize":"1","localStorage-contentId":"px01-bkt-0001/content/countall/e6/e6e29e35a58d31b050dca8aadbbae9d93cb0fe55/17d56aced5601c3"}
x-amz-meta-pointerkey: px01-bkt-0001/pointer/countall/e6/e6e29e35a58d31b050dca8aadbbae9d93cb0fe55
Date: Thu, 25 Nov 2021 10:37:43 GMT
Authorization: AWS s3proxy:YsqkJSjI0VOlGHUTXbY64ySGDCk=
User-Agent: jclouds/2.4.0 httpclient/4.1.1
Content-Length: 1
Content-Type: text/plain;charset=UTF-8
Host: 127.0.0.1:39941
Connection: keep-alive
Accept-Encoding: gzip,deflate
```

Only diff I can spot:
* object path
* size is 1 vs 3116 bytes
* content type "text/plain;charset=UTF-8" vs "application/atom+xml; charset=utf-8; type=packages"
* w/ 1.9.0 s3proxy tried removing "suspicious" things (like `$` from header value and change path `/.metadata` to without dot in it, same issue, this request failed always with 1.9.0 (while works okay with 1.8.0)

**Comment by gaul** [comment]
Since this happens in your tests could you try git bisecting the changes between 1.8.0 and 1.9.0?  Specifically I wonder if upgrading from Jetty 9.2 to 9.4 caused this issue.  4686ede0bdfd546f6167a872f57bb879c6bc5275 has a suspicious test change that matches the differing `Content-Type` that you describe.

**Comment by cstamas** [comment]
@gaul I think I found the issue. The Jetty upgrade (from 9.2 to 9.4) brought in some interesting things along the lines :smile: 

Jetty 9.4.2x+ by default "normalizes" the request (as in HttpParser as in HttpGenerator). As I see (by analyzing headers what our app sends to s3proxy), they "see" two different things: 
* our app (jclouds) sends header `text/plain;charset=utf-8` (lower case), I can confirm this from from gist of "server side" (link above).
* but it seems that s3proxy gets "normalized" headers from Jetty? It becomes `text/plain;charset=UTF-8`.

Due shading, could not verify my claim yet, as it seems setting `-Dorg.gaul.shaded.org.eclipse.jetty.http.HttpParser.STRICT=true` on CLI for s3proxy did not cut it, so am still looking.

**Comment by cstamas** [comment]
Yes, https://github.com/gaul/s3proxy/commit/4686ede0bdfd546f6167a872f57bb879c6bc5275#diff-154f5886142935454ec888f1b6e3d0f39091f3255506bf495e5d871f8e0099efL637 is exactly what am seeing

**Comment by gaul** [comment]
b47e8387fd4c6dd727398b4177b063661590c8b6 sets Jetty `STRICT` mode so we should be seeing the original value.  Maybe Jetty 9.4 changes this behavior?

**Comment by cstamas** [comment]
@gaul hm, I did not spot that, but seems is not working: I was looking for WARNING in log that Jetty should emit when this is set... but is not.

**Comment by cstamas** [comment]
Here is the log that should be triggered by STRICT mode: https://github.com/eclipse/jetty.project/blob/jetty-9.4.41.v20210516/jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java#L245
I also see it in decompiled code:
![screenshot](https://user-images.githubusercontent.com/45165/143447589-353db456-add4-47d7-b831-fa9dbbd03269.png)

**Comment by cstamas** [comment]
As I see (decompiled) s3proxy code, this property you (and I tried) to set is totally unused, due how Jetty is constructed in S3Proxy class....

**Comment by gaul** [comment]
S3Proxy was using an older Jetty for many years to maintain compatibility with Java 7.  The 9.4 upgrade may have changed how this property is set.  System-wide properties are a bad mechanism anyway...

**Comment by cstamas** [comment]
Tried updating Jetty to latest (9.4.44.v20210927)  but same issue still happens (and UT fails as well).

**Comment by janbartel** [comment]
@gaul you don't have to use the system property to set the compliance mode. 

Here: https://github.com/gaul/s3proxy/blob/master/src/main/java/org/gaul/s3proxy/S3Proxy.java#L91  change your constructor to `new HttpConnectionFactory(new HttpConfiguration(), HttpCompliance.LEGACY);`.

**Comment by gaul** [comment]
@janbartel thank you for the hint!  I fixed this in gaul/s3proxy@2e61c3862656419f2a0452d9ceab22be311e84fb.

**Comment by gaul** [comment]
Jetty 9.4 and newer still have mysterious behavior removing spaces from Content-Type, changing from `"text/html; charset=UTF-8"` to `"text/html;charset=utf-8"`, but I think this issue is addressed.

**Comment by joakime** [comment]
Many values in various common headers come from lookups (called a `Trie` in Jetty code) to reduce object creation and GC load.
Typically the first time a header appears it's nuances with OWS (Optional White Space) and case sensitivity are cached in a Trie.
For `Content-Type`, it has it's own Trie maintained in the MimeTypes object, and is pre-loaded with various common MimeTypes.

If you notice, the OWS was removed, and the charset was normalized from the spec unrecommended `UTF-8` to spec recommended `utf-8`.
That's because for `text/html` with `utf-8` it's a pre-cached value from `MimeTypes`.

https://github.com/eclipse/jetty.project/blob/jetty-9.4.44.v20210927/jetty-http/src/main/java/org/eclipse/jetty/http/MimeTypes.java#L68

When RFC7230 and RFC7231 were ratified (back in June 2014) is when this conformance change to MimeTypes was made into Jetty.

**Comment by cstamas** [comment]
Thanks for detailed explanation.

**Comment by cstamas** [comment]
@gaul can we expect a patch release (like 1.9.1) with this fix? As current 1.9.0 is broken due this issue with Mime types that are handled as explained in @joakime comment above (like text/plain is).

**Comment by gaul** [comment]
#410.
